### PR TITLE
Do not centralize type/cname lookups through inet_db

### PR DESCRIPTION
**Pull request by josevalim**
Before this patch, inet_res:getbyname lookups
were centralized through the inet_db process, which
made it a bottleneck on concurrent machines that need
to establish many connections to several hosts quickly.

Since the lookups are now concurrent, the update of
resolve.conf has been changed to use atomic operations
that replace the list of servers and list of names.

The inet_db API has also been cleaned up to no longer
expose operations that delete all searches/namespaces/etc.
Those functions seem safe to remove as they are not
used anywhere within Erlang/OTP. The atomic replace
operations can be made public if desired in the future.
There are also other unused functions which we can
remove in the future.

This is a regen of #2818 and follow up to #2891.

### CODE REVIEW

**Comment by josevalim** [comment]
@sneako if by any chance you could give this branch a try and provide us feedback, it would be fantastic! :heart:

**Comment by sneako** [comment]
I tested out this PR, along with #2891 and unfortunately we saw some instability in our cluster as load increased. This instability was accompanied by many `eaddrinuse` errors in our logs.

This morning, I deployed #2891 on its own to test it in isolation. I let it run in production, on a single node, for over an hour under what we consider to be very high load, and it performed well with no issues at all. Therefore it is possible that there is an issue with this PR, and further testing should be done before it is merged. 

However, I am not 100% certain that the instabilities we saw in our application were caused solely by this PR, and we have made several significant improvements to the application in the meantime, so I would like to run another test soon.

In summary, I think we should hold off on merging this PR for now, but I will report back again after my next test which I plan to run very soon.

**Comment by RaimoNiskanen** [comment]
Since practically the only change in this PR is to change cache lookups from server calls to ETS lookups, it seems unlikely that it could cause `eaddrinuse` errors almost anywhere.  So I hope you will find out that it was some other change that was the culprit...

**Comment by RaimoNiskanen** [comment]
@josevalim: Should we maybe think about caching not only gethostbyname and gethostbyaddr lookups, but all record types?  I was under the impression that `inet_res` mainly was used when you have to look up other than IN,A, IN,AAAA and IN,PTR records. (?).

**Comment by RaimoNiskanen** [comment]
A question to these and all users of `inet_res`: is it used on the public Internet or only in "safe" environments?  And why are you using `inet_res` instead of the native resolver (`inet_gethost_native` from `inet`)?

**Comment by sneako** [comment]
Right, I agree that this PR probably did not cause the eaddrinuse errors.

In my case, we are using `inet_res` instead of `inet_gethost_native` because we found `inet_gethost_native` became a bottleneck in our application frequently causing message queue pile ups. We are running in Kubernetes, and found `inet_res` to be much faster. Since we switched to `inet_res` and started running [node local dns pods](https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/), we stopped experiencing these message pile ups.

For example, here is the difference we see on our Kubernetes pods:
```
iex(app@host)1> for _ <- 1..50, do: fn -> :inet_gethost_native.gethostbyname('erlang.org') end |> :timer.tc |> elem(0)
[452, 739, 522, 439, 424, 324, 365, 308, 271, 303, 321, 490, 373, 309, 326, 528,
 392, 294, 365, 334, 286, 319, 332, 397, 295, 293, 275, 260, 372, 348, 317, 272,
 351, 277, 479, 442, 387, 419, 804, 372, 302, 386, 360, 590, 415, 300, 287, 538,
 389, 1788]

iex(app@host)2> for _ <- 1..50, do: fn -> :inet_res.gethostbyname('erlang.org') end |> :timer.tc |> elem(0)
[55, 28, 38, 53, 51, 20, 26, 27, 88, 46, 26, 12, 20, 34, 14, 35, 39, 84, 27, 34,
 20, 20, 29, 28, 46, 69, 46, 44, 20, 11, 37, 33, 21, 18, 50, 36, 19, 46, 89, 39,
 43, 38, 39, 32, 32, 59, 32, 43, 34, 58]
```

**Comment by RaimoNiskanen** [comment]
@sneako: Well, I'll be darned! For speed reasons!  In that benchmark you will get almost all answers served from the `inet_db` cache, but I guess `inet_res` indeed improves actual performance.  Maybe also largely thanks to the cache?

So this means that you ignore the platform's resolver configuration and instead ask DNS. (?)
Are you using the `inet` framework https://erlang.org/doc/apps/erts/inet_cfg.html and set the lookup order to e.g `[file, dns]`, or are you calling inet_res directly?

I was thinking about introducing caching in the `inet` framework instead of only for `inet_res` so `inet_gethost_native` could benefit from caching.

**Comment by sneako** [comment]
Hehe yes, we have a need for speed 😄  We are using an inet_cfg file with `{lookup, [dns, native]}.` set in order to use `inet_res`, and do not call it directly.

I definitely like the idea of adding caching for `inet_gethost_native` so that you can simply rely on the system dns, but still avoid the syscalls when a result is cached.

**Comment by sneako** [comment]
I am running another test of this PR right now. So far things are looking good and it seems that our other recent changes have fixed the `eaddrinuse` issue we saw last time. I'm going to let this run for a while a keep increasing the load on this test node, and report back. In the meantime, checkout how fast `inet_res` is now that we skip the gen_server call:

```
> for _ <- 1..50, do: fn -> :inet_res.gethostbyname('erlang.org') end |> :timer.tc |> elem(0)
[17, 8, 27, 6, 10, 7, 14, 4, 3, 4, 3, 3, 3, 3, 4, 6, 14, 12, 8, 6, 7, 4, 4, 18,
 4, 10, 5, 5, 5, 4, 6, 4, 4, 5, 9, 4, 4, 3, 12, 4, 4, 4, 3, 3, 4, 6, 3, 4, 4, 3]
```

**Comment by sneako** [comment]
I have been running this patch in production now for almost 2 hours on one node in our cluster. Using weighted routing, I have been sending extra traffic to this node, keeping the schedulers around 90% saturated, and performance has been very good! It appears that our other recent changes have fixed the `eaddrinuse` errors we saw last time we tried out this patch.

**Comment by RaimoNiskanen** [comment]
Gosh!  Mean time went down from 38 µs to 6 µs.  That is fast!

> I definitely like the idea of adding caching for inet_gethost_native so that you can simply rely on the system dns, but still avoid the syscalls when a result is cached.

I was thinking about adding caching to `inet` for all resolver backends.  I hope it would not be too large a change...

**Comment by RaimoNiskanen** [comment]
Merged.  Thank you for your pull request!
